File: LAB/Lab_2/Z1.scala
def indexSwap[A](i: Int, j: Int)(list: List[A]): List[A] = 
    def swapIndicesHelper(j: Int, list: List[A], v: A): (A, List[A]) =
        (j, list) match
            case (0, hd :: tl) => (hd, v :: tl)
            case (_, hd :: tl) =>
                val (x, y) = swapIndicesHelper(j - 1, tl, v)
                (x, hd :: y)
            case _ => (v, Nil)
    
    def swapIndices(i: Int, j: Int, list: List[A]): List[A] =
        (i, j, list) match
            case (0, _, hd :: tl) =>
                val (x, y) = swapIndicesHelper(j - 1, tl, hd)
                x :: y
            case (_, 0, hd :: tl) =>
                val (x, y) = swapIndicesHelper(i - 1, tl, hd)
                x :: y
            case (_, _, hd :: tl) => hd :: swapIndices(i - 1, j - 1, tl)
            case _ => list

    swapIndices(i, j, list) 

@main def main() =
    val list = List(0,1,2,3,4,5)

    val indexSwap15 = indexSwap(1, 5)
    val indexSwap51 = indexSwap(5, 1)
    val indexSwap9999 = indexSwap(99, 99)
    val indexSwap099 = indexSwap(0, 99)

    println(indexSwap15(list))
    println(indexSwap15(List(5,4,3,2,1,0)))
    println(indexSwap51(list))
    println(indexSwap9999(list))
    println(indexSwap099(list))
        

File: LAB/Lab_2/Z1.ml
let indexSwap i j =
  let rec swapIndicesHelper j list v =
    match (j, list) with
    | (0, hd :: tl) -> (hd, v :: tl)
    | (_, hd :: tl) -> let (x, y) = swapIndicesHelper (j - 1) tl v in (x, hd :: y) 
    | _ -> (v, [])
  in

  let rec swapIndices i j list =
    match (i, j, list) with
    | (0, _, hd :: tl) -> let (x, y) = swapIndicesHelper (j - 1) tl hd in x :: y
    | (_, 0, hd :: tl) -> let (x, y) = swapIndicesHelper (i - 1) tl hd in x :: y
    | (_, _, hd :: tl) -> hd :: swapIndices (i - 1) (j - 1) tl
    | _ -> list
  in

  fun list -> swapIndices i j list
;;

let indexSwap15 = indexSwap 1 5;;
indexSwap15 [0;1;2;3;4;5];;
indexSwap15 [0;1;2];;
let indexSwap51 = indexSwap 5 1;;
indexSwap51 [0;1;2;3;4;5];;
let indexSwap9999 = indexSwap 99 99;;
indexSwap9999 [0;1;2;3];;